lib/posix/pthread.c

```diff
@@ -216,18 +216,18 @@ int pthread_create(pthread_t *newthread, const pthread_attr_t *attr,
  */
 int pthread_setcancelstate(int state, int *oldstate)
 {
+	k_spinlock_key_t cancel_key;
 	struct posix_thread *pthread = to_posix_thread(pthread_self());
 
 	if (state != PTHREAD_CANCEL_ENABLE &&
 	    state != PTHREAD_CANCEL_DISABLE) {
 		return EINVAL;
 	}
 
+	cancel_key = k_spin_lock(&pthread->cancel_lock);
 	*oldstate = pthread->cancel_state;
```

Comment:
This looks like it's fixing a potential race here -- the old version was fetching `cancel_state` without protection of a lock.

Reply:
I think so as well